clowder/kafka.go

```diff
@@ -16,60 +16,86 @@ package clowder
 
 import (
 	"fmt"
+
 	"github.com/RedHatInsights/insights-operator-utils/kafka"
 	api "github.com/redhatinsights/app-common-go/pkg/api/v1"
 )
 
 // Common constants used for logging and error reporting
 const (
-	noBrokerConfig = "warning: no broker configurations found in clowder config"
-	noSaslConfig   = "warning: SASL configuration is missing"
-	noTopicMapping = "warning: no kafka mapping found for topic %s"
+	noOriginalBroker = "warning: no original broker configuration found; aborting"
+	noBrokerConfig   = "warning: no broker configurations found in clowder config"
+	noSaslConfig     = "warning: SASL configuration is missing"
+	noTopicMapping   = "warning: no kafka mapping found for topic %s"
 )
 
 // UseBrokerConfig tries to replace parts of the BrokerConfiguration with the values
-// loaded by Clowder
-func UseBrokerConfig(brokerCfg *kafka.BrokerConfiguration, loadedConfig *api.AppConfig) {
-	// make sure broker(s) are configured in Clowder
+// loaded by Clowder. It expects brokerCfg to already be initialized
+func UseBrokerConfig(brokerCfg kafka.BrokersConfig, loadedConfig *api.AppConfig) kafka.BrokersConfig {
+	numBrokerConfigs := len(brokerCfg)
+	if numBrokerConfigs == 0 {
+		// if original brokers config is totally empty, do nothing.
+		// this shouldn't happen, but we need to control this scenario
+		// to avoid panics.
+		fmt.Println(noOriginalBroker)
+		return brokerCfg
+	}
 	if loadedConfig.Kafka != nil && len(loadedConfig.Kafka.Brokers) > 0 {
-		broker := loadedConfig.Kafka.Brokers[0]
-		// port can be empty in api, so taking it into account
-		if broker.Port != nil {
-			brokerCfg.Address = fmt.Sprintf("%s:%d", broker.Hostname, *broker.Port)
-		} else {
-			brokerCfg.Address = broker.Hostname
+		numClowderBrokers := len(loadedConfig.Kafka.Brokers)
+		// if original config has fewer brokers than clowder's, we append additional
+		// brokerConfiguration items with topic, clientId, and group from existing
+		// items, and the rest will be filled with data from clowder's brokers.
+		// When appending, it's most probable that a new slice is returned due to
+		// original capacity not being enough, which is why the brokerCfg slice is
+		// returned
+		for len(brokerCfg) < numClowderBrokers {
+			brokerCfg = append(brokerCfg, &kafka.BrokerConfiguration{
+				Topic:    (brokerCfg)[numBrokerConfigs-1].Topic,
+				ClientID: (brokerCfg)[numBrokerConfigs-1].ClientID,
+				Group:    (brokerCfg)[numBrokerConfigs-1].Group,
+				// Since this will have data from Clowder's loadedConfig, always enable
+				Enabled: true,
+			})
```

Comment:
I don't understand this. The Clowder configuration should replace the loaded configuration, not appending a few brokers at the end of them

Reply:
This is for the case when the clowder configuration has more brokers than what we originally configured. From my point of view, we are  interested in those additional brokers (the backups). But if we are not, then this code can basically go away

Reply:
I think that if we have both broker configured by our service configuration and using Clowder, we should  use just Clowder ones and "forget" about the service configured ones.

Most of the time we are just using the config.toml, which usually contains local Kafka endpoints, and Clowder is over-seeding it

Reply:
TBH I did think about it at some point, but didn't implement it to not change the entire logic, but that does make 100% sense.

I'll open the discussion in a team standup, and we can decide what to do about it

Reply:
Well, we're good then.

@joselsegura based on our discussion, this PR does what is expected:
- in these lines we are adding more "BrokerConfiguration" items to the original slice of configs
- then in lines 60 to 87, we replace all the `BrokerConfiguration` items' content with Clowder's `Kafka.Brokers` content

Reply:
Ok, I didn't realize about that